### ISSUE TITLE
Find a way to prevent people from writing synchronous C# code

### ISSUE THREAD

**Issue by davidebbo** [Azure/azure-webjobs-sdk-script#774]
We ran into a couple cases (one [here](https://social.msdn.microsoft.com/Forums/azure/en-US/5c142d90-adef-4d7b-8a30-3224b6229d93/app-could-be-having-an-intermediat-issue-or-failing-to-start?forum=AzureFunctions)) where user's apps deadlocked because they were making synchronous outgoing http calls, e.g.

    HttpResponseMessage response = client.PostAsync(url , httpContent).Result;

That causes the thread to be blocked while waiting for the response. Once you throw enough traffic at it, all the threads are blocked, and there are none left to get scheduled and do the work, so you have a deadlock. See http://stackoverflow.com/questions/10343632/httpclient-getasync-never-returns-when-using-await-async.

@lindydonna suggested that we could potentially use Roslyn's analyzer to detect the situation and treat it as an error. Though it's an aggressive move, we really want to make it very hard for people to do this.

**Comment by fabiocav** [comment]
This could indeed be a case of thread exhaustion and slow ramp up, severely impacting responses, but we should not be seeing a deadlock with the pattern above as we explicitly prevent the ASP.NET synchronization context from being captured and flowing down to the function. We have tests in place to cover that scenario [here](https://github.com/Azure/azure-webjobs-sdk-script/blob/1ab54a94916c8180105af1e43b24a483f070a53f/test/WebJobs.Script.Tests/TestScripts/FunctionGeneration/HttpTrigger-CSharp/run.csx).

So the issue is a bit different than what is described in the SO post above (which would happen in a single threaded sync context like ASP.NET, WPF, WinForms, etc.), where we're not deadlocking, but holding on to thread pool threads for much longer than needed, leading to exhaustion and slowing processing (with enough load, likely to te point where requests are timing out)

We have discussed preventing that pattern in functions (including through the use of analyzers) before, but I agree with what you say about it being pretty aggressive. If we decide to do this, we need to do it carefully to make sure we're not breaking valid scenarios. 

Where you able to repro the above issue in one of your apps?

**Comment by davidebbo** [comment]
@ahmelsayed tried but I'm not sure if he did. However, for the msdn thread above, we analyzed the dump and saw over 500 threads mostly all blocking on the http call, so it looked very much like a deadlock.

**Comment by fabiocav** [comment]
Interesting. Would be good to investigate as we should not have a condition where we'd schedule continuations back on the request thread. The deadlock scenario is reproducible with a single request (no load required)

**Comment by davidebbo** [comment]
Here is what the stacks look like in that dump (I just renamed the function):

```
>	mscorlib.dll!System.Threading.Monitor.Wait(object obj, int millisecondsTimeout, bool exitContext)	Unknown
 	mscorlib.dll!System.Threading.Monitor.Wait(object obj, int millisecondsTimeout)	Unknown
 	mscorlib.dll!System.Threading.ManualResetEventSlim.Wait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.SpinThenBlockingWait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task.InternalWait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task<System.Net.Http.HttpResponseMessage>.GetResultCore(bool waitCompletionNotification)	Unknown
 	mscorlib.dll!System.Threading.Tasks.Task<System.__Canon>.Result.get()	Unknown
 	ƒ-SomeFunctionMethodName#R*9ad986bd-bd95-48be-a512-efbc6f3318ba#3-0!Submission#0.PostToPowerBI(object realTimeEvent, Microsoft.Azure.WebJobs.Host.TraceWriter log)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start<System.__Canon>(ref System.__Canon stateMachine)	Unknown
 	ƒ-SomeFunctionMethodName#R*9ad986bd-bd95-48be-a512-efbc6f3318ba#3-0!Submission#0.PostToPowerBI(object realTimeEvent, Microsoft.Azure.WebJobs.Host.TraceWriter log)	Unknown
 	ƒ-SomeFunctionMethodName#R*9ad986bd-bd95-48be-a512-efbc6f3318ba#3-0!Submission#0.Run(string eventHubMessage, Microsoft.Azure.WebJobs.Host.TraceWriter log)	Unknown
 	mscorlib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start<System.__Canon>(ref System.__Canon stateMachine)	Unknown
 	ƒ-SomeFunctionMethodName#R*9ad986bd-bd95-48be-a512-efbc6f3318ba#3-0!Submission#0.Run(string eventHubMessage, Microsoft.Azure.WebJobs.Host.TraceWriter log)	Unknown
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	Microsoft.Azure.WebJobs.Script.dll!Microsoft.Azure.WebJobs.Script.Description.DotNetFunctionInvoker.InvokeCore(object[] parameters, Microsoft.Azure.WebJobs.Script.Description.FunctionInvocationContext context)	Unknown
```

**Comment by fabiocav** [comment]
Will see if I can repro this, and agree we need to make these situations harder to run into.

I think you guys are right on the solution. Just think that what we're seeing is that once we exhaust thread pool treads we're seeing work queuing up waiting for threads to become available (since growing is slow) and leading to the behavior we're seeing.

I'm assuming that under low load, this function completes as expected (which wouldn't happen with the usual deadlock scenario, described in the SO post), but once threads become scarce because of the blocking pattern used, we see things grind to a halt.

**Comment by davidebbo** [comment]
Yes, I think that's correct that it takes load to reach that situation. But once reached, it's not just slow, it's a true deadlock.

**Comment by fabiocav** [comment]
I see. I guess hitting the max number of threads and getting to the point where the pool no longer grows, with enough queued items, would explain this (does that make sense to you?).

I agree, though, that the bottom line is that we need to make this harder to happen. The analyzer approach should work pretty well. I actually have a prototype of an analyzer to do something similar, from when we were talking about using them to prevent this and *fire-and-forget* scenarios.

**Comment by davidebbo** [comment]
One more data point is that the host status API was timing out when in that condition. So no more http requests could be served.

Danger of preventing all async is that we'll break too many things. Some of our templates aren't even async! In practice, it's mostly an issue when making outgoing http calls, but that's really hard to detect reliably.

Let's chat about it tomorrow.

**Comment by fabiocav** [comment]
Sounds good. Yeah, it makes sense that all APIs would be dead at that point.

**Comment by fabiocav** [comment]
In this scenario, timeouts would also be busted, as that logic would also never have a chance to run.  :(

**Comment by lindydonna** [comment]
@fabiocav Are you referring to the host timeout?  That does indeed sound bad.

**Comment by ahmelsayed** [comment]
Yes I think I was able to repro this. I created 2 function apps. 
the first one had an http trigger function with just 

```csharp
using System.Net;

public static async Task<HttpResponseMessage> Run(HttpRequestMessage req, TraceWriter log)
{
    await Task.Delay(TimeSpan.FromMinutes(2));
    return req.CreateResponse(HttpStatusCode.OK);
}
```

and the other function app had 1 queue triggered function

```csharp
using System;

public static void Run(string myQueueItem, TraceWriter log)
{
    using (var client = new HttpClient())
    {
        var response = client.GetAsync("theOtherFunctionUrl").Result;
        log.Info(response.StatusCode.ToString());
    }
}
```

I had this custom host.json though

```json
{
    "queues": {
      "batchSize": 32,
      "newBatchThreshold": 2000
    }
}
```

preloaded the queue with 5k messages, then enabled the function. I think it's more of the process is running out of threadpool threads and can't schedule work anymore. You can actually go to kudu's process explorer and see the number of threads rising in the process

**Comment by fabiocav** [comment]
Yeah, @ahmelsayed, that's my thought as well. Once we reach max threads and enough queued work items for the thread pool, we're done.

**Comment by ctolkien** [comment]
This is a little concerning. We've a scenario where we need to do some long running process to which there is no `async` option, we'll be blocking a thread until the results come back. I'm not sure what the solution is here?

Aside (and I don't mean to dissuade the work you guys do), but this isn't really a "serverless" solution when you need to consider how many threads you can run... on your server.

**Comment by davidebbo** [comment]
@ctolkien we won't do anything drastic that would block such scenarios. If anything, we'll target specific scenarios that have no reason not to be async. e.g. anytime a method that returns a Task has a `.Result` instead of an `await`, it really screams to be made async.

**Comment by ctolkien** [comment]
OK great, however am I likely going to run into deadlock issues in this scenario as well if a lot of requests are thrown at my function (that's my primary issue, not so much around guidance on using async where possible)?

**Comment by davidebbo** [comment]
No, you should be fine if you just do heavyweight synchronous work. The deadlock only applies to the pattern where you start an async op and then call `.Result` on it.

**Comment by ctolkien** [comment]
Ahhhh.. OK, no cause for concern then 👍

**Comment by lindydonna** [comment]
@ctolkien Yes, the thing that's easy to forget is that .Result is essentially doing a busy-wait on a lock of another thread. Holding a lock creates the potential for deadlock, whereas your synchronous APIs wouldn't have this problem.

**Comment by lindydonna** [comment]
We're seeing customers run into this and open support tickets, so we should investigate a fix.

**Comment by ganesp** [comment]
@fabiocav 
I am seeing the deadlock problem in one of the applications, where we make huge number of calls to this method which downloads a blob. And if you notice in the stack trace below, there is .Result usage.

Only a small percentage of calls are seeing deadlock. 
I suppose, making this call await would resolve the deadlock problem.

It took me a while to figure out why the deadlock is intermittent and very rare.

```
System.Threading.ManualResetEventSlim.Wait(Int32, System.Threading.CancellationToken)
System.Threading.Tasks.Task.SpinThenBlockingWait(Int32, System.Threading.CancellationToken)
System.Threading.Tasks.Task.InternalWait(Int32, System.Threading.CancellationToken)
System.Threading.Tasks.Task`1[[System.__Canon, mscorlib]].GetResultCore(Boolean)
Caller -->  A method in my application with .Result usage on a async method which downloads a blob from storage (a HTTP call)
```
